### ISSUE TITLE
Exceptions within threads are silently discarded when testing

### ISSUE THREAD

**Issue by paulbutcher** [flix/flix#5123]
The following test passes 😳

```
@Test
def testBug(): Unit \ IO = region r {
    spawn { bug!("Something bad happened") } @ r;
    Thread.sleep(Time/Duration.fromSeconds(1))
}
```

**Comment by paulbutcher** [comment]
Thinking about this, this is a larger problem than just what happens during testing. We need to think about what is the right thing to do if an unhandled exception happens within a thread in any Flix program.

The simplest thing to do would be to register a [default uncaught exception handler](https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/lang/Thread.html#setDefaultUncaughtExceptionHandler(java.lang.Thread.UncaughtExceptionHandler)) which reports the exception and then [terminates the JVM](https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/lang/Runtime.html#halt(int))

But perhaps we need to provide a mechanism by which a Flix programmer can modify this behaviour or register a handler of their own?

My temptation is to do the simple thing right now, but could be persuaded that we need to do something more complex.

Thoughts?

**Comment by JonathanStarup** [comment]
I think we should just crash hard yeah, any known exceptions would come from java interop and should be handled and translated into a Result or something like that. You could also `spawn { try .. .catch { case e => Channel.send(Err(e))}` if you really wanted to handle the any exceptions in the thread

**Comment by magnus-madsen** [comment]
> The simplest thing to do would be to register a [default uncaught exception handler](https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/lang/Thread.html#setDefaultUncaughtExceptionHandler(java.lang.Thread.UncaughtExceptionHandler)) which reports the exception and then [terminates the JVM](https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/lang/Runtime.html#halt(int))

While harsh, I would vote for this. 

This will certainty aid both us and developers in ensuring that errors are not lost (which I consider a worse outcome).

**Comment by magnus-madsen** [comment]
- [ ] Print stack trace and exit the JVM if a thread dies (by installing a uncaught exception handler in Region class).

**Comment by paulbutcher** [comment]
I seem to have hit a Catch-22. The [redirection of stdout and stderr](https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/tools/Tester.scala#L216) during testing means that if I call `halt` for an uncaught exception, the messages are silently swallowed.

But... If I call `exit` to allow these messages to be output with a shutdown hook then everything deadlocks because the JVM tries to unwind itself cleanly, which it can't.

@magnus-madsen I'm not sure I understand what these redirects get us. Can you enlighten me?

**Comment by magnus-madsen** [comment]
> I seem to have hit a Catch-22. The [redirection of stdout and stderr](https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/tools/Tester.scala#L216) during testing means that if I call `halt` for an uncaught exception, the messages are silently swallowed.
> 
> But... If I call `exit` to allow these messages to be output with a shutdown hook then everything deadlocks because the JVM tries to unwind itself cleanly, which it can't.
> 
> @magnus-madsen I'm not sure I understand what these redirects get us. Can you enlighten me?

The redirect are necessary for the test infrastructure to work, but I forget the details.

**Comment by magnus-madsen** [comment]
Is it right that today if a thread throws it is just silently ignored?

Would setting an exception handler that at least calls printStackTrace be a step forward?

**Comment by paulbutcher** [comment]
> Would setting an exception handler that at least calls printStackTrace be a step forward?

It would be if the redirection within the test framework didn't silently swallow it.

**Comment by magnus-madsen** [comment]
> 

I think its because the test framework collects the prints to show them at the end, not in the middle of the run. You can try to experiment with this.

**Comment by paulbutcher** [comment]
It would help a great deal if I knew why it was redirecting output at all?

**Comment by magnus-madsen** [comment]
> It would help a great deal if I knew why it was redirecting output at all?

Yes, its to ensure that while running the unit tests the output is never messed up.

It will always print 

- test01
- test02

and so forth. No matter what a test prints.

**Comment by magnus-madsen** [comment]
Then at the end (if there was errors (??) it will print the stdout/stderr

**Comment by paulbutcher** [comment]
Is this a case where the cure is worse than the disease? Most tests should output nothing whatsoever, surely?

**Comment by magnus-madsen** [comment]
> Is this a case where the cure is worse than the disease? Most tests should output nothing whatsoever, surely?

I don't know, there was a lot of complains about the previous system. Maybe we just need to always print the captured output at the end? Would that work?

https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/tools/Tester.scala#L109

Is it because a test is SUCCESSFUL, but there is a thread that crashes?

**Comment by paulbutcher** [comment]
That would be an improvement, certainly. But:

1. It wouldn't help at all for cases where the thread deadlocks (which will happen much more frequently now that we have an implied `join` at the end of every region).
2. It still allows tests to pass which have unhandled exceptions (which is definitely wrong)
3. We would have a conflict between the behaviour we want when testing and the behaviour we want for production code

**Comment by magnus-madsen** [comment]
Perhaps when a thread fails, the owner thread should also fail. (Implementing that is not so easy), but conceptually would that make sense and solve all our problems?

**Comment by paulbutcher** [comment]
> Perhaps when a thread fails, the owner thread should also fail. (Implementing that is not so easy), but conceptually would that make sense and solve all our problems?

I think, sadly, provably impossible on the JVM.

**Comment by magnus-madsen** [comment]
> It wouldn't help at all for cases where the thread deadlocks (which will happen much more frequently now that we have an implied join at the end of every region).

True, but here I assume the test framework itself would hang, so that would at least not be bad.

If I understood correctly the  real problem is that a non-main thread dies in a test case and we never learn about it!

**Comment by magnus-madsen** [comment]
> > Perhaps when a thread fails, the owner thread should also fail. (Implementing that is not so easy), but conceptually would that make sense and solve all our problems?
> 
> I think, sadly, provably impossible on the JVM.

I think Java 21 (or whatever) will enable this. But I am not sure how we would implement it in the meantime.

We could have an exception handler and signal failure on a result channel?

**Comment by magnus-madsen** [comment]
I have a simple pragmatic proposal for now:

Change https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/tools/Tester.scala#L109 to also print STDERR for SUCCESSFUL tests. Then have the UncaughtHandler ensure that we print to STDERR. I think this is pragmatic.

**Comment by magnus-madsen** [comment]
And requires very little extra work now.

**Comment by paulbutcher** [comment]
We could write something that could work *some* of the time. But this is an area where Java is well known to be deficient:

![Uploading Screenshot 2022-12-17 at 14.55.20.png…]()

**Comment by paulbutcher** [comment]
> I have a simple pragmatic proposal for now:
> 
> Change https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/tools/Tester.scala#L109 to also print STDERR for SUCCESSFUL tests. Then have the UncaughtHandler ensure that we print to STDERR. I think this is pragmatic.

It would be an acceptable solution for testing. But not for production code (which we *really* want to die if there's an uncaught exception).

**Comment by magnus-madsen** [comment]
> It would be an acceptable solution for testing. But not for production code (which we really want to die if there's an uncaught exception).

But the Java default is not to die, right? The thread dies, prints its stacktrace and that is all. I think that is what we get today already, no?

In any case, I think Java 21 or whatever will help us immensely.

**Comment by magnus-madsen** [comment]
(Its only the tester that redirect output; if that was not clear)

**Comment by paulbutcher** [comment]
I'd love to understand the complaints about test output: I wonder if there might be an alternative solution to whatever the problem is which avoids all this nastiness?

**Comment by magnus-madsen** [comment]
> I'd love to understand the complaints about test output: I wonder if there might be an alternative solution to whatever the problem is which avoids all this nastiness?

The point is that garbage output of tests should not mess up the summary printed to the terminal.

**Comment by paulbutcher** [comment]
Why are tests outputting garbage though?

**Comment by paulbutcher** [comment]
I guess I can *maybe* see a justification for redirecting stdout. But why on Earth would I ever not want to see something that's output to stderr?

**Comment by magnus-madsen** [comment]
Hence; let us just output stderr for any test no matter whether it failed or succeeded. 

We could even fail a succeeding test if it outputted to stderr.

Again, I think this is pragmatic and requires very little code 👍

**Comment by paulbutcher** [comment]
OK, let me have a play with that and see where it gets us.

**Comment by paulbutcher** [comment]
The simplest solution (if we're being pragmatic) would be to simply not redirect stderr. This wouldn't allow us to fail tests which write to stderr, but would cope with the deadlock case (which I really do care about). Would that work for you?

**Comment by magnus-madsen** [comment]
> The simplest solution (if we're being pragmatic) would be to simply not redirect stderr. This wouldn't allow us to fail tests which write to stderr, but would cope with the deadlock case (which I really do care about). Would that work for you?

No, that's a step back on why we introduced redirection in the first place. Sorry! :)

**Comment by paulbutcher** [comment]
Can you (or someone) then please explain to me why we have tests that are outputting to stderr as a matter of course? What's the valid usecase for that?

**Comment by paulbutcher** [comment]
What we seem to be saying is that we can choose between:

1. Code that fails in a clear and understandable way, or
2. Output from tests that's pretty in the event that the test outputs to stderr

I would choose "clear and understandable" over "pretty" every day of the week and twice on Sundays.

**Comment by magnus-madsen** [comment]
> Can you (or someone) then please explain to me why we have tests that are outputting to stderr as a matter of course? What's the valid usecase for that?

I think the point is rather that when things fail (maybe 20 tests fail with lots of output (could be stack traces)) then you still want a nice summary and then at the end to have all errors collected in one place.

**Comment by magnus-madsen** [comment]
This is not something we invented, e.g. lots of JS frameworks do it (maybe that is a bad argument...)

But anyway, this ticket was not intended to be a design discussion ;)

**Comment by magnus-madsen** [comment]
(By the way, I do feel your pain in discovering this. I just think it was a very bad bug in the Test framework, but I think it is very fixable and will not come back to bite...)

**Comment by paulbutcher** [comment]
I'm sorry, I think we have to have a design discussion.

What we're currently proposing is that it's OK to have a test framework in which a test can throw an exception, output to stderr, deadlock, and as a result never output anything at all.

I don't think that that's acceptable. At all.

**Comment by paulbutcher** [comment]
> (By the way, I do feel your pain in discovering this. I just think it was a very bad bug in the Test framework, but I think it is very fixable and will not come back to bite...)

But it's a bug which isn't fixed by the currently proposed solution.

**Comment by magnus-madsen** [comment]
Why is it not fixed? My proposal is that writing to stderr will both cause the test to fail and print it's output (but at the end of the test run?)

**Comment by paulbutcher** [comment]
Because if the test deadlocks, the end of the test run is never.

**Comment by magnus-madsen** [comment]
I see. What normally happens when deadlocks happen? Does java print?

**Comment by paulbutcher** [comment]
Nothing at all happens, sadly.

**Comment by paulbutcher** [comment]
I guess we could work around this by:

1. Having some kind of timeout and writing whatever we have in our stderr/stdout after that timeout?
2. Running each test in its own thread (which I suspect comes with some performance overhead) and timing out and moving on to the next test?

**Comment by paulbutcher** [comment]
Not sure I like either. But I like them much more than withholding information from the programmer (which is what our current proposal does)

**Comment by magnus-madsen** [comment]
Could the test framework install it's own uncaught handler and complain loudly that something died or is stuck for a test?

**Comment by paulbutcher** [comment]
How would that be any different from just outputting stderr immediately?

**Comment by magnus-madsen** [comment]
It would only output for thread death and deadlock immediately

**Comment by paulbutcher** [comment]
Right. But I believe I'm right in saying that pretty much every case where there's any output to stderr results from an  unhandled exception?

**Comment by paulbutcher** [comment]
Oh! Unless what you mean is that we treat unhandled exceptions in the "main" thread differently from unhandled exceptions in threads created by `spawn`?

That might be do-able, although it'll be fiddly.

**Comment by paulbutcher** [comment]
> Oh! Unless what you mean is that we treat unhandled exceptions in the "main" thread differently from unhandled exceptions in threads created by spawn?

@magnus-madsen can you confirm that this is what you meant?

**Comment by magnus-madsen** [comment]
I made a fix in https://github.com/flix/flix/pull/5136. You can try it out.

**Comment by paulbutcher** [comment]
> I made a fix in #5136. You can try it out.

This does exactly what I expected: It's fine, but it silently discards the exception information in the event of deadlock. Given:

```
def explode(): Unit \ IO = 
    println("Boom");
    bug!("Something bad happend")

@Test 
def test01(): Unit \ IO = region r {
    let (tx, rx) = Channel.unbuffered(r);
    spawn {
        explode();
        Channel.send((), tx)
    } @ r;
    spawn {
        Channel.recv(rx)
    } @ r
}
```

The output is:

```
Running 1 tests...

   TEST  test01
```

There is no way to see the exception information.

This is *very important*. Debugging concurrency bugs is difficult enough without the test framework maliciously hiding information from you. And deadlock is a *very common* failure mode.

**Comment by magnus-madsen** [comment]
This is an important example, but I want to approach it not from the perspective of the test framework, but from the language.

I think we thread 1 crashes, we should rethrow the exception in the main thread. 
(This is also what StructuredTaskScope will do eventually). 

Thus we either exit the region normally knowing that all threads terminated normally or we re-throw.

I think this would work as follows:

Region needs to store a collection of throw exceptions (or maybe just the first).
When a thread dies, it needs to interrupt its parent thread (not yet sure how to find which one).
Upon region we now have to check for interruption / InterruptedException.
If there is an exception we rethrow it. 

Thus I want the usual property of exceptions (but across threads): Either you encounter a handler or **everything dies in main**.

**Comment by paulbutcher** [comment]
OK, that would solve the problem, but I'll have to have a think about whether it's implementable. I'm not sure that it is, but let me have a think.

**Comment by magnus-madsen** [comment]
I mean the example was great! I had not thought of that issue.

But I am also really happy with the solution, if it can work. It really feels like how exceptions "should have worked".

**Comment by paulbutcher** [comment]
> I mean the example was great! I had not thought of that issue.

Apologies for failing to explain it clearly: that's been exactly the example I've had in mind throughout this conversation (which I guess is why we've been failing to communicate).

**Comment by magnus-madsen** [comment]
👍

**Comment by paulbutcher** [comment]
So I've looked at this and it's definitely impossible to implement with certainty.

There is no way to throw an exception in one thread from another. The closest we can get is [`Thread.interrupt`](https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/lang/Thread.html#interrupt()), but this will only work if:

1. The thread is blocked on something that's interruptible (e.g. `Thread.join` or [`Lock.lockInterruptibly`](https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/util/concurrent/locks/Lock.html#lockInterruptibly())
2. The thread checks [`Thread.isInterrupted`](https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/lang/Thread.html#isInterrupted()).

In particular, it won't work if:

1. The thread is blocked on something that isn't interruptible (e.g. any `synchronized` method, [`Lock.lock`](https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/util/concurrent/locks/Lock.html#lock()))
2. The thread is CPU bound without checking `Thread.isInterrupted` occasionally.

This means that we're hosed if:

1. The "main" thread calls any Java interop method (which can clearly do anything whatsoever).
2. Is in a busy loop.

The best we can do is to make sure that any time we block in Flix generated or library code, we block interruptibly, and then hope that the main thread will block eventually. And that user code doesn't catch `InterruptedException`.

This *might* be enough. Perhaps.

**Comment by magnus-madsen** [comment]
Maybe I am confused, but I thought the idea would be to put this logic in Region.exit. Thus once we are about to leave the region (all threads have stopped) we check if any crashes and rethrow.

**Comment by paulbutcher** [comment]
That assumes that we *get* to `Region.exit`, which we aren't guaranteed to.

We won't if the main thread calls some Java method via interop which blocks. And we won't if the main thread is CPU bound. Or if the main thread is blocked (prior to `Region.exit`) on something that it's expecting from the spawned thread but which it will never receive because that thread has crashed.

**Comment by paulbutcher** [comment]
There's definitely nothing we can do about the first case (Java interop is completely out of our control).

I guess the second (CPU bound) is unlikely enough that we *might* be able to ignore it (although it makes me feel icky).

The third we can control by ensuring that we always block interruptibly, and I guess we can avoid the problem of the programmer catching `InterruptedException` by catching it ourselves before user code gets a look-in.

**Comment by paulbutcher** [comment]
To make this concrete: here is a slightly modified version of my earlier example which deadlocks in a way which means that `Region.exit` is never called:

```
def explode(): Unit \ IO = 
    println("Boom");
    bug!("Something bad happend")

@Test 
def test01(): Unit \ IO = region r {
    let (tx, rx) = Channel.unbuffered(r);
    spawn {
        explode();
        Channel.send((), tx)
    } @ r;
    Channel.recv(rx)
}
```

**Comment by magnus-madsen** [comment]
If the spawning thread (e.g. `main`) enters an infinite loop there is nothing to be done. 

If it waits on a channel, I think interrupting it can work.

In any case, we just need something until we have `StructuredTaskScope`.

Can we start by ensuring that crashes are registered in `Region` and rethrown in `exit`?

**Comment by magnus-madsen** [comment]
Other than infinite loops, it seems `StructuredTaskScope` solves many of these issues.

(I am not proposing that we try to replicate that; rather we should just have something that is good enough until then).

**Comment by paulbutcher** [comment]
As far as I can see from the `StructuredTaskScope` documentation, it doesn't re-throw exceptions until after the child threads have been `join`ed. So I don't think it helps particularly.

But I agree that interrupting can be made to work for the case where the main thread is blocked on something we control (i.e. a channel). We're out of luck if it's blocked on something via Java interop or is in an infinite loop. I dislike the fact that we will potentially be hiding valuable information from the programmer in these cases, but they're going to be rare (much rarer than "normal" deadlock) so I can live with it.

I still think the tradeoff between "pretty" test results and "correct" test results should really be made in favour of "correct" every time. But as always I'll defer to your judgement.

**Comment by magnus-madsen** [comment]
> We're out of luck if it's blocked on something via Java interop or is in an infinite loop. 

I believe, but I could be wrong, that Project Loom will fix everything related to Java blocking (as long as it is not CPU bound), i.e. I think all IO will more or less become interruptible. (Of course infinite loops will hang).

I'll think more carefully about the test framework. At least, it is not for the runtime, only for tests.

One point I completely forgot to make was I was trying to achieve separation between *running a test* and *showing/collecting test results*. I was partially interested in this to support the Test API in VSCode at some point in the future. Now maybe redirecting stdout/stderr as part of that design is a terrible idea, but it was what I found at the time (and saw other frameworks doing).

**Comment by magnus-madsen** [comment]
I think in our implementation we will only be able to rethrow once all threads have terminated or crashes (which is OK, but not the best). I believe this:


> [ShutdownOnFailure](https://download.java.net/java/early_access/loom/docs/api/jdk.incubator.concurrent/jdk/incubator/concurrent/StructuredTaskScope.ShutdownOnFailure.html) captures the first exception and shuts down the task scope. This class is intended for cases where the results of all subtasks are required ("invoke all"); if any subtask fails then the results of other unfinished subtasks are no longer needed. If defines methods to throw an exception if any of the subtasks fail.


from `StructuredTaskScope` is much stronger and ultimately what we want.

**Comment by paulbutcher** [comment]
Actually, the comment about VSCode makes me realise that there is a way that we could both have our cake and eat it:

We could make use of terminal control codes to effectively create two different regions of output: one containing test results, the other containing test output (imagine two "windows": one with prettily formatted test results, the other stdout/stderr). This would enable us to output test stdout and stderr immediately and still format things however we wanted to.

It would be lots of work, mind.

**Comment by paulbutcher** [comment]
> I think in our implementation we will only be able to rethrow once all threads have terminated or crashed (which is OK, but not the best). I believe this:

That won't give us a solution to the deadlock case though, will it? Which is critical for us to solve well (I know I'm a stuck record on this, for which apologies, but I have *many* scars from trying to debug concurrency bugs which are leading me to dig my heels in)

**Comment by magnus-madsen** [comment]
What's the deadlock example that breaks (assuming use of StructuredTaskScope or similar)?

In the last example, I would assume that explode destroys its thread which in turn interrupts the main thread that is waiting for a message where the exception is then rethrown. I dont know if we can make it work, but with `StructuredTaskScope` that should definitely happen.

**Comment by magnus-madsen** [comment]
... because the primitives we use to implement channels can all be interrupted (AFAIK).

**Comment by paulbutcher** [comment]
> because the primitives we use to implement channels can all be interrupted (AFAIK).

Yes, this is what I meant when I said:

> I agree that interrupting can be made to work for the case where the main thread is blocked on something we control (i.e. a channel).

But you then went on to say:

> I think in our implementation we will only be able to rethrow once all threads have terminated or crashed

Re throwing only once all threads have terminated or crashed definitely isn't good enough.

**Comment by paulbutcher** [comment]
I think I've got enough of a picture to be able to put together an implementation that might work. I'll do so and run it past you.

**Comment by magnus-madsen** [comment]
What if we, in addition, add a timeout of say 10s after which we both (1) mark the test as TIMEOUT (or whatever) (2) immediately (or later, since we have marked it failed) dump the stdout/stderr?

**Comment by paulbutcher** [comment]
> What if we, in addition, add a timeout of say 10s after which we both (1) mark the test as TIMEOUT (or whatever) (2) immediately (or later, since we have marked it failed) dump the stdout/stderr?

Yup, I suggested that above:

> Having some kind of timeout and writing whatever we have in our stderr/stdout after that timeout?

But I think we're talking at cross purposes. We don't, in fact, have to wait for all threads to crash or terminate.

**Comment by paulbutcher** [comment]
So, here's what I propose:

We generate a new `UncaughtExceptionHandler` class as follows:

```
public class UncaughtExceptionHandler implements Thread.UncaughtExceptionHandler {
  Region r;
  UncaughtExceptionHandler(Region r) { this.r = r; }
  public void uncaughtException(Thread t, Throwable e) { r.reportChildException(e); }  
}
```

We modify the generated `Region` class as follows:

```
public class Region {
  private ConcurrentLinkedQueue<Thread> threads = new ConcurrentLinkedQueue<Thread>();

  private Thread parentThread = Thread.currentThread();
  private volatile Throwable childException = null;

  final public void reportChildException(Throwable e) {
    childException = e;
    parentThread.interrupt();
  }

  final public void reThrowChildException() throws Throwable {
    if (childException != null)
      throw childException;
  }

  final public void spawn(Runnable r) {
    Thread t = new Thread(r);
    t.setUncaughtExceptionHandler(new UncaughtExceptionHandler(this));
    t.start();
    threads.add(t);
  }

  final public void exit() throws InterruptedException {
    Thread t;
    while ((t = threads.poll()) != null)
      t.join();
  }
}
```

And we modify the code generated for `Scope` so it takes the following form:

```
    Region r = new Region();
    try {
      << Body of scope >>
      r.exit();
    } finally {
      r.reThrowChildException();
    }
```

@JonathanStarup: I'd be grateful for your thoughts.

**Comment by paulbutcher** [comment]
Possible enhancement, perhaps we want to create a new type of exception `ChildThreadException` or similar which wraps the re-thrown exception (to make it clear that the exception hasn't come from the thread it's being re-thrown in).

**Comment by paulbutcher** [comment]
It's worth pointing out again that this is a lot of work, and a significant change to the exception behaviour that developers are likely to expect, just to get pretty test results.

**Comment by paulbutcher** [comment]
Implementation note: I'm proposing to make `childException` volatile, which may not be strictly necessary: interrupting one thread by another is a synchronisation event, as is `join`. But this is a bit of "belt and braces" because I'm not 100% sure how exceptions might affect that (they probably don't, but there's no harm making it volatile).

**Comment by paulbutcher** [comment]
@JonathanStarup my proposal won't interact nicely with this code within Channels:

https://github.com/flix/flix/blob/master/main/src/library/Concurrent/Channel.flix#L338-L339

Was there a particular issue you were trying to address here?

**Comment by JonathanStarup** [comment]
We wanted deadlocked "flix threads" to be interruptable but I don't remember it being a thought out thing, especially since it should also be done for the other operations to achieve that goal

**Comment by paulbutcher** [comment]
> We wanted deadlocked "flix threads" to be interruptable but I don't remember it being a thought out thing, especially since it should also be done for the other operations to achieve that goal

That's a good point: for this to work at all reliably (it will never work 100% reliably for reasons already discussed) we should do an audit of every instance we call a Java function in the standard library and ensure that we're using something interruptible.

**Comment by paulbutcher** [comment]
Yes, I considered this. I *think* I'm right in saying (although I haven't traced the codepath all the way through) that we do this for every continuation. Which would mean that adding the try/catch block here would add a try/catch overhead to every continuation (which I suspect is more overhead than we would prefer?)

**Comment by magnus-madsen** [comment]
> Yes, I considered this. I _think_ I'm right in saying (although I haven't traced the codepath all the way through) that we do this for every continuation. Which would mean that adding the try/catch block here would add a try/catch overhead to every continuation (which I suspect is more overhead than we would prefer?)

I see. I *think* that is the case. @JonathanStarup ? :)

**Comment by JonathanStarup** [comment]
in the normal case we call `invoke` or `unwind` based on tail position. `run` is just implemented using `unwind`. `run` is only used by spawn

**Comment by paulbutcher** [comment]
So I've come back to this after a hiatus. I'd like to check that we still think that the proposal [above](https://github.com/flix/flix/issues/5123#issuecomment-1357932909) is the direction we'd like to take. For completeness, this will have the following implications:

* The changes to the code generated for regions as described above
* Modify the channel code to lock interruptibly instead of uninterruptibly
* Modify the `compileForceMethod` in `GenLazyClasses.scala` to avoid the use of `ACC_SYNCHRONIZED` (which is not interruptible)
* Audit all remaining code to ensure that we never block uninterruptibly (the two examples above are the ones I'm aware of, but there may be others)
* Avoid any use of non-interruptible locking in any future code (not sure how we can police/enforce this?)
* Whatever we do, it is not guaranteed to work 100% reliably (if code called via Java interop blocks uninterruptibly).

**Comment by magnus-madsen** [comment]
Sounds good. What do you say @JonathanStarup?

**Comment by magnus-madsen** [comment]
How would you implement `compileForceMethod`?

**Comment by paulbutcher** [comment]
> How would you implement `compileForceMethod`?

I haven't looked in detail yet, but probably switch to an explicit `ReentrantLock` and call its `lockInterruptibly` method

**Comment by paulbutcher** [comment]
(bottom line is that we can *never* use `ACC_SYNCHRONIZED` in any code we generate)

**Comment by JonathanStarup** [comment]
I'm pretty sure that we don't do concurrency stuff outside of Lazy and Channel. I remember some difficulty implementing force without `ACC_SYNCHRONIZED` (without details) but if you have the time, maybe you could write the lazy class @paulbutcher and I can write it to the backend?

> Avoid any use of non-interruptible locking in any future code

I guess the best we can do is not expose the unwanted methods in the flix API

**Comment by magnus-madsen** [comment]
Do we want a new ticket that clarifies exactly what kind of locks we should be using?

**Comment by paulbutcher** [comment]
> Do we want a new ticket that clarifies exactly what kind of locks we should be using?

Unfortunately it's slightly more complicated than which types of locks to use: it's also which method you use to lock (e.g. `RentrantLock` has both `lock` (which can't be interrupted) and `lockInterruptibly` (which can). Similarly `Condition` has both `await` (which can be interrupted) and `awaitUninterruptibly` (which can't).

I'm not sure that an additional ticket would add much value (I'm 90% of the way through an initial PR for this, although there will be a little more to do). The problem will be ensuring that future code does the right thing (and I'm not sure that a ticket will do a lot to help with that?)

**Comment by magnus-madsen** [comment]
Maybe something in the `docs` folder?

**Comment by magnus-madsen** [comment]
And a grep through the code for ACC_SYNC or what is it

**Comment by paulbutcher** [comment]
> And a grep through the code for ACC_SYNC or what is it

Yup, that's how I identified `compileForceMethod` as needing modification.

**Comment by paulbutcher** [comment]
> Maybe something in the `docs` folder?

That makes sense. Let me see what I can come up with.

Better would be something that automatically detects/forbids this kind of thing, but it's not obvious how to do that.
